analysis.ml

```diff
@@ -1,34 +1,44 @@
 open Instr
 
-let successors (instrs : instruction_stream) pc =
+let pcs (instrs : instruction_stream) : pc array =
+  Array.mapi (fun pc _ -> pc) instrs
+
+let successors_at (instrs : instruction_stream) pc : pc list =
   let pc' = pc + 1 in
-  let next = if pc' = Array.length instrs then [] else [pc'] in
+  let instr = instrs.(pc) in
   let resolve = Instr.resolve instrs in
-  match instrs.(pc) with
-  | Decl_const _
-  | Decl_mut _
-  | Assign _
-  | Drop _
-  | Clear _
-  | Read _
-  | Label _
-  | Comment _
-  | Osr _
-  | Print _ -> next
+  match instr with
+  | Decl_const _ | Decl_mut _ | Assign _ | Drop _ | Clear _ | Read _
+  | Label _ | Comment _ | Osr _ | Print _ ->
+    let is_last = pc' = Array.length instrs in
+    if is_last then [] else [pc']
+  | Stop -> []
   | Goto l -> [resolve l]
+  | Branch (_e, l1, l2) when l1 = l2 -> [resolve l1]
   | Branch (_e, l1, l2) -> [resolve l1; resolve l2]
-  | Stop -> []
 
-let predecessors (instrs : instruction_stream) =
-  let preds = Array.map (fun _ -> []) instrs in
-  let mark_successor pc pc' =
-    preds.(pc') <- pc :: preds.(pc') in
-  for pc = 0 to Array.length instrs - 1 do
-    List.iter (mark_successor pc) (successors instrs pc)
-  done;
-  preds
+let successors (instrs : instruction_stream) : pc list array =
+  let succs_at pc = successors_at instrs pc in
+  Array.map succs_at (pcs instrs)
+
+let predecessors (instrs : instruction_stream) : pc list array =
```

Comment:
For most programming language features and style there are several stage to learning, one of them being "finding out how/when to use it" and a later one "finding out how/when not to use it". You seem to be somewhere in the middle :-) Your code for this function is exactly as long as the previous one, but it is harder to understand (more higher-order functions) and quadratic instead of linear.